synapse/storage/databases/main/stream.py

```diff
@@ -68,8 +70,12 @@
 
 
 def generate_pagination_where_clause(
-    direction, column_names, from_token, to_token, engine
-):
+    direction: str,
+    column_names: Tuple[str, str],
+    from_token: Optional[Tuple[int, int]],
+    to_token: Optional[Tuple[int, int]],
+    engine,
```

Comment:
Can we change this to `Union[PostgresEngine, SqliteEngine]`? It feels a bit wrong to leave it untyped

Reply:
👍 I think I was afraid this would create crazy import loops.

Reply:
Actually, it seems there's a `BaseDatabaseEngine` which I should use instead.